File: Python/BubbleSort/BJAlgo_1838_suc.py
# 가장 많이 이동한 (index 값이 변경된) 횟수 = 정렬이 완료되기까지 반복한 횟수
from sys import stdin


class Pair:
    def __init__(self, index, value):
        self.index = index
        self.value = value


N = int(stdin.readline())
data = list(map(int, stdin.readline().split()))

pq = list()

for i in range(len(data)):
    pq.append(Pair(i, data[i]))

pq.sort(key=lambda obj: obj.value)

result = 0
for i in range(len(pq)):
    tmp = pq[i]
    if result < (tmp.index - i):
        result = tmp.index - i

print(result)

File: Java/BubbleSort/BJAlgo_1838_suc.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.PriorityQueue;

public class BJAlgo_1838_suc {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());
        PriorityQueue<Pair> data = new PriorityQueue<>();

        String[] inputs = br.readLine().split(" ");
        for (int i = 0; i < N; i++) {
            data.add(new Pair(i, Integer.parseInt(inputs[i])));
        }

        int result = 0;
        int index = 0;
        while (!data.isEmpty()) {
            Pair tmp = data.poll();
            if (result < (tmp.index - index)) {
                result = tmp.index - index;
            }

            index++;
        }

        System.out.println(result);
    }
}

class Pair implements Comparable<Pair>{
    int index;
    int value;

    Pair(int index, int value) {
        this.index = index;
        this.value = value;
    }

    @Override
    public int compareTo(Pair pair) {
        return (int) (this.value - pair.value);
    }
}
